Transaction recovery handling finalizer does not permit wildflyserver to be terminated when there is no active pod

Finalizer makes the operator removal being stuck when there is no active pod under operator management. The operator repeats error message[1] and never shuts down.
The activity may end up with panic.

```
{"level":"info","ts":1568704825.2354705,"logger":"controller_wildflyserver","msg":"WildflyServer is marked for deletion. Waiting for finalizers to clean the workspace","Request.Namespace":"default","Request.Name":"quickstart"}
{"level":"info","ts":1568704825.2425647,"logger":"controller_wildflyserver","msg":"Transaction recovery scaledown processing","Request.Namespace":"default","Request.Name":"quickstart","Pod Name":"quickstart-0","IP Address":"172.17.0.6"}
{"level":"info","ts":1568704825.2728443,"logger":"controller_wildflyserver","msg":"Enabling recovery listener for processing scaledown at quickstart-0","Request.Namespace":"default","Request.Name":"quickstart"}
{"level":"error","ts":1568704829.2799802,"logger":"controller_wildflyserver","msg":"Verify if operator JBOSS_HOME variable determines the place where the application server is installed","Request.Namespace":"default","Request.Name":"quickstart","quickstart.JBOSS_HOME":"/opt/jboss/wildfly","error":"Cannot execute JBoss CLI command /opt/jboss/wildfly/bin/jboss-cli.sh --output-json -c --command='/subsystem=transactions:write-attribute(name=recovery-listener, value=true)' at pod quickstart-0. Cause: cannot execute '/opt/jboss/wildfly/bin/jboss-cli.sh --output-json -c --command='/subsystem=transactions:write-attribute(name=recovery-listener, value=true)'': command terminated with exit code 1","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/jmesnil/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/wildfly/wildfly-operator/pkg/controller/wildflyserver.(*ReconcileWildFlyServer).checkRecovery\n\t/Users/jmesnil/Developer/wildfly-operator/pkg/controller/wildflyserver/wildflyserver_controller.go:968\ngithub.com/wildfly/wildfly-operator/pkg/controller/wildflyserver.(*ReconcileWildFlyServer).processTransactionRecoveryScaleDown.func1\n\t/Users/jmesnil/Developer/wildfly-operator/pkg/controller/wildflyserver/wildflyserver_controller.go:887"}
{"level":"error","ts":1568704829.2823055,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"wildflyserver-controller","request":"default/quickstart","error":"Finalizer processing: failed transaction recovery for WildflyServer default:quickstart name Error: Found 1 errors:\n [[Cannot enable transaction recovery listener for scaling down pod quickstart-0, error: Cannot execute JBoss CLI command /opt/jboss/wildfly/bin/jboss-cli.sh --output-json -c --command='/subsystem=transactions:write-attribute(name=recovery-listener, value=true)' at pod quickstart-0. Cause: cannot execute '/opt/jboss/wildfly/bin/jboss-cli.sh --output-json -c --command='/subsystem=transactions:write-attribute(name=recovery-listener, value=true)'': command terminated with exit code 1]],","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/jmesnil/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/jmesnil/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.12/pkg/internal/controller/controller.go:217\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/jmesnil/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.12/pkg/internal/controller/controller.go:158\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/jmesnil/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/jmesnil/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:134\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/jmesnil/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:88"}

{"level":"info","ts":1568705107.4781928,"logger":"controller_wildflyserver","msg":"Reconciling WildFlyServer","Request.Namespace":"default","Request.Name":"quickstart"}
{"level":"info","ts":1568705107.483508,"logger":"controller_wildflyserver","msg":"WildflyServer is marked for deletion. Waiting for finalizers to clean the workspace","Request.Namespace":"default","Request.Name":"quickstart"}
{"level":"info","ts":1568705107.485902,"logger":"controller_wildflyserver","msg":"Transaction recovery scaledown processing","Request.Namespace":"default","Request.Name":"quickstart","Pod Name":"quickstart-0","IP Address":"172.17.0.6"}
{"level":"info","ts":1568705107.512193,"logger":"controller_wildflyserver","msg":"Enabling recovery listener for processing scaledown at quickstart-0","Request.Namespace":"default","Request.Name":"quickstart"}
{"level":"info","ts":1568705110.7030013,"logger":"controller_wildflyserver","msg":"Reloading WildFly for recovery listener being activated at pod quickstart-0","Request.Namespace":"default","Request.Name":"quickstart"}
{"level":"info","ts":1568705113.2886152,"logger":"controller_wildflyserver","msg":"Query to find the transaction recovery port to force scan at pod quickstart-0","Request.Namespace":"default","Request.Name":"quickstart"}
panic: assignment to entry in nil map

goroutine 378 [running]:
github.com/wildfly/wildfly-operator/pkg/controller/wildflyserver.(*ReconcileWildFlyServer).checkRecovery(0xc0004e5dd0, 0x1624960, 0xc000ac82e0, 0xc00027ea80, 0xc00091a1c0, 0x4, 0xc0000576e0, 0xc000057710, 0xebad1f, 0xc000611110)
        /Users/jmesnil/Developer/wildfly-operator/pkg/controller/wildflyserver/wildflyserver_controller.go:995 +0xa6e
github.com/wildfly/wildfly-operator/pkg/controller/wildflyserver.(*ReconcileWildFlyServer).processTransactionRecoveryScaleDown.func1(0xc0006a20e0, 0xc00027ea80, 0xc000aca300, 0xc000aca330, 0xc00091a1c0, 0x1624960, 0xc000ac82e0, 0xc0004e5dd0)
        /Users/jmesnil/Developer/wildfly-operator/pkg/controller/wildflyserver/wildflyserver_controller.go:887 +0x2e9
created by github.com/wildfly/wildfly-operator/pkg/controller/wildflyserver.(*ReconcileWildFlyServer).processTransactionRecoveryScaleDown
        /Users/jmesnil/Developer/wildfly-operator/pkg/controller/wildflyserver/wildflyserver_controller.go:866 +0x35c
```

--- a/go.mod
+++ b/go.mod
@@ -42,6 +42,7 @@ require (
 	github.com/sirupsen/logrus v1.4.1 // indirect
 	github.com/spf13/pflag v1.0.3
 	github.com/stretchr/testify v1.3.0
+	github.com/tevino/abool v0.0.0-20170917061928-9b9efcf221b5
 	go.uber.org/atomic v1.3.2 // indirect
 	go.uber.org/multierr v1.1.0 // indirect
 	go.uber.org/zap v1.9.1 // indirect

--- a/pkg/controller/util/go_utilities.go
+++ b/pkg/controller/util/go_utilities.go
@@ -72,3 +72,18 @@ func SortPodListByName(podList *corev1.PodList) *corev1.PodList {
 	})
 	return podList
 }
+
+// MapMerge merges the two maps together and returns the result.
+// If one of them is nil then is ommitted from the merged result
+// If both maps are null then an empty initialized map is returned
+// The second map rewrites data of the first map in the result if there are such
+func MapMerge(firstMap map[string]string, secondOverwritingMap map[string]string) map[string]string {
+	returnedMap := make(map[string]string)
+	for v, k := range firstMap {
+		returnedMap[v] = k
+	}
+	for v, k := range secondOverwritingMap {
+		returnedMap[v] = k
+	}
+	return returnedMap
+}

--- a/pkg/controller/util/remote_ops.go
+++ b/pkg/controller/util/remote_ops.go
@@ -27,6 +27,8 @@ var (
 	timestampFromLogLine = regexp.MustCompile(`^[0-9]{4}-[0-9]{2}-[0-9]{2}[^ ]+`)
 	// start of the time timestamp
 	startOfTheTimeTimestamp = time.Unix(0, 0)
+	// socket dial timeout
+	socketDialTimeout = 10 * time.Second
 )
 
 //ExecRemote executes a command inside the remote pod
@@ -87,7 +89,7 @@ func ExecRemote(pod *corev1.Pod, command string) (string, error) {
 func SocketConnect(hostname string, port int32, command string) (string, error) {
 	// connect to socket
 	toConnectTo := fmt.Sprintf("%v:%v", hostname, port)
-	conn, err := net.Dial("tcp", toConnectTo)
+	conn, err := net.DialTimeout("tcp", toConnectTo, socketDialTimeout)
 	if err != nil {
 		return "", fmt.Errorf("Cannot process TCP connection to %v, error: %v",
 			toConnectTo, err)

--- a/pkg/controller/util/wildfly_mgmt.go
+++ b/pkg/controller/util/wildfly_mgmt.go
@@ -22,8 +22,8 @@ var (
 	MgmtOpReload = ":reload()"
 	// MgmtOpRestart is a JBoss CLI command for restarting WFLY server
 	MgmtOpRestart = ":shutdown(restart=true)"
-	// MgmtOpTxnEnableRecoveryListener is a JBoss CLI command for enabling txn recovery listener
-	MgmtOpTxnEnableRecoveryListener = "/subsystem=transactions:write-attribute(name=recovery-listener, value=true)"
+	// MgmtOpTxnCheckRecoveryListener is a JBoss CLI command for enabling txn recovery listener
+	MgmtOpTxnCheckRecoveryListener = "/subsystem=transactions:read-attribute(name=recovery-listener)"
 	// MgmtOpTxnProbe is a JBoss CLI command for probing transaction log store
 	MgmtOpTxnProbe = "/subsystem=transactions/log-store=log-store:probe()"
 	// MgmtOpTxnRead is a JBoss CLI command for reading transaction log store

--- a/pkg/controller/util/wildfly_mgmt.go
+++ b/pkg/controller/util/wildfly_mgmt.go
@@ -202,7 +202,7 @@ func ExecuteOpAndWaitForServerBeingReady(mgmtOp string, pod *corev1.Pod, jbossHo
 		return false, fmt.Errorf("Cannot run operation '%v' at application container for down pod %s, error: %v", mgmtOp, podName, err)
 	}
 	if !IsMgmtOutcomeSuccesful(jsonResult) {
-		return false, fmt.Errorf("Not succefully runnin management operation '%v' for pod %s. JSON output: %v",
+		return false, fmt.Errorf("Not succefully running management operation '%v' for pod %s. JSON output: %v",
 			mgmtOp, podName, jsonResult)
 	}
 	for serverStateCheckCounter := 0; err != nil && serverStateCheckCounter < reloadRetryCount; serverStateCheckCounter++ {

--- a/pkg/controller/wildflyserver/wildflyserver_controller.go
+++ b/pkg/controller/wildflyserver/wildflyserver_controller.go
@@ -11,6 +11,7 @@ import (
 	"sync"
 
 	"github.com/go-logr/logr"
+	"github.com/tevino/abool"
 	wildflyv1alpha1 "github.com/wildfly/wildfly-operator/pkg/apis/wildfly/v1alpha1"
 	wildflyutil "github.com/wildfly/wildfly-operator/pkg/controller/util"
 
@@ -180,13 +181,17 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 			//   this will be retried at next reconciliation.
 			r.recorder.Event(wildflyServer, corev1.EventTypeNormal, "WildFlyServerDeletion",
 				"Scaledown processing started before removal. Consult operator log for status.")
-			if err := r.finalizeWildflyServer(reqLogger, wildflyServer); err != nil {
+			requeue, err := r.finalizeWildflyServer(reqLogger, wildflyServer)
+			if err != nil {
 				return reconcile.Result{}, err
+			}
+			if requeue {
+				return reconcile.Result{Requeue: true}, nil
 			}
 
 			// Remove WildflyServer. Once all finalizers have been removed, the object will be deleted.
 			wildflyServer.SetFinalizers(wildflyutil.RemoveFromList(wildflyServer.GetFinalizers(), wildflyServerFinalizer))
-			err := r.client.Update(context.TODO(), wildflyServer)
+			err = r.client.Update(context.TODO(), wildflyServer)
 			if err != nil {
 				return reconcile.Result{}, err
 			}
@@ -236,12 +241,12 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 	}
 	// Processing recovery on pods which are planned to be removed because of scale down is in progress now
 	mustReconcile, err := r.processTransactionRecoveryScaleDown(reqLogger, wildflyServer, int(numberOfPodsToScaleDown), podList)
+	if mustReconcile { // server state was updated (or/and some erro could happen), we need to reconcile
+		return reconcile.Result{Requeue: true}, err
+	}
 	if err != nil {
 		reqLogger.Error(err, "Failures during scaling down recovery processing", "Desired replica size", wildflyServerSpecSize,
 			"Number of pods to be removed", numberOfPodsToScaleDown)
-	}
-	if mustReconcile { // server state was updated, we need to reconcile
-		return reconcile.Result{Requeue: true}, err
 	}
 
 	mustReconcile, requeue, err := r.checkStatefulSet(wildflyServer, foundStatefulSet, podList)
@@ -270,7 +275,6 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 		reqLogger.Error(err, "Failed to get LoadBalancer %s", loadBalancerName)
 		return reconcile.Result{}, err
 	}
-
 	// Check if the headless service already exists, if not create a new one
 	foundHeadlessService := &corev1.Service{}
 	headlessServiceName := headlessServiceName(wildflyServer)
@@ -351,8 +355,7 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 
 	if updateWildflyServer {
 		if err := r.client.Status().Update(context.Background(), wildflyServer); err != nil {
-			reqLogger.Error(err, "Failed to update WildFlyServer status.")
-			return reconcile.Result{}, err
+			return reconcile.Result{}, fmt.Errorf("Failed to update WildFlyServer status. Error: %v", err)
 		}
 		requeue = true
 	}
@@ -405,17 +408,17 @@ func (r *ReconcileWildFlyServer) checkStatefulSet(wildflyServer *wildflyv1alpha1
 		// Scaling down statefulset to number of pods that were cleaned by recovery
 		calculatedStatefulSetReplicaSize := int32(numberOfPods - numberOfPodsToShutdown)
 		desiredStatefulSetReplicaSize = calculatedStatefulSetReplicaSize
-		foundStatefulSet.Spec.Replicas = &desiredStatefulSetReplicaSize
 		if wildflyServerSpecSize <= calculatedStatefulSetReplicaSize && *foundStatefulSet.Spec.Replicas > calculatedStatefulSetReplicaSize {
 			log.Info("Scaling down and updating replica size to "+strconv.Itoa(int(calculatedStatefulSetReplicaSize)),
 				"StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
+			foundStatefulSet.Spec.Replicas = &desiredStatefulSetReplicaSize
 			update = true
 		}
 		// There are some unclean pods which can't be scaled down
 		if wildflyServerSpecSize < calculatedStatefulSetReplicaSize {
 			log.Info("Statefulset was not fully scaled to the desired replica size "+strconv.Itoa(int(wildflyServerSpecSize))+
-				" while StatefulSet is to be at size "+strconv.Itoa(int(calculatedStatefulSetReplicaSize))+
-				". Some pods were not cleaned by recovery. Verify status of the WildflyServer "+wildflyServer.Name,
+				" (current StatefulSet size: "+strconv.Itoa(int(calculatedStatefulSetReplicaSize))+
+				"). Some pods were not cleaned by recovery. Verify status of the WildflyServer "+wildflyServer.Name,
 				"StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
 			r.recorder.Event(wildflyServer, corev1.EventTypeWarning, "WildFlyServerScaledown",
 				"Transaction recovery slowed down the scaledown.")
@@ -754,27 +757,30 @@ func (r *ReconcileWildFlyServer) routeForWildFly(w *wildflyv1alpha1.WildFlyServe
 	return route
 }
 
-func (r *ReconcileWildFlyServer) finalizeWildflyServer(reqLogger logr.Logger, w *wildflyv1alpha1.WildFlyServer) error {
+func (r *ReconcileWildFlyServer) finalizeWildflyServer(reqLogger logr.Logger, w *wildflyv1alpha1.WildFlyServer) (bool, error) {
 	podList, err := r.getPodsForWildFly(w)
 	if err != nil {
-		return fmt.Errorf("Finalizer processing: failed to list pods for WildflyServer %v:%v name Error: %v", w.Namespace, w.Name, err)
+		return false, fmt.Errorf("Finalizer processing: failed to list pods for WildflyServer %v:%v name Error: %v", w.Namespace, w.Name, err)
 	}
-	_, err = r.processTransactionRecoveryScaleDown(reqLogger, w, len(podList.Items), podList)
+	requeue, err := r.processTransactionRecoveryScaleDown(reqLogger, w, len(podList.Items), podList)
 	if err != nil {
 		// error during processing transaction recovery, error from finalizer
-		return fmt.Errorf("Finalizer processing: failed transaction recovery for WildflyServer %v:%v name Error: %v", w.Namespace, w.Name, err)
+		return false, fmt.Errorf("Finalizer processing: failed transaction recovery for WildflyServer %v:%v name Error: %v", w.Namespace, w.Name, err)
+	}
+	if requeue {
+		return true, nil
 	}
 	for _, v := range w.Status.Pods {
 		if v.State == wildflyv1alpha1.PodStateScalingDownRecoveryInvestigation {
-			return fmt.Errorf("Finalizer processing: recovery in progress at %v/%v at pod %v", w.Namespace, w.Name, v.Name)
+			return false, fmt.Errorf("Finalizer processing: recovery in progress at %v/%v at pod %v", w.Namespace, w.Name, v.Name)
 		}
 		if v.State != wildflyv1alpha1.PodStateScalingDownClean {
-			return fmt.Errorf("Finalizer processing: transaction recovery processed with unfinished transactions at %v/%v at pod %v with state %v",
+			return false, fmt.Errorf("Finalizer processing: transaction recovery processed with unfinished transactions at %v/%v at pod %v with state %v",
 				w.Namespace, w.Name, v.Name, v.State)
 		}
 	}
 	reqLogger.Info("Finalizer finished succesfully", "WildflyServer Namespace", w.Namespace, "WildflyServer Name", w.Name)
-	return nil
+	return false, nil
 }
 
 func (r *ReconcileWildFlyServer) addWildflyServerFinalizer(reqLogger logr.Logger, w *wildflyv1alpha1.WildFlyServer) error {
@@ -835,7 +841,7 @@ func (r *ReconcileWildFlyServer) processTransactionRecoveryScaleDown(reqLogger l
 	scaleDownErrors := sync.Map{}     // errors occured during processing the scaledown for the pods
 
 	// Setting-up the pod status - status is used to decide if the pod could be scaled (aka. removed from the statefulset)
-	var updated bool
+	updated := abool.New()
 	for scaleDownIndex := 1; scaleDownIndex <= numberOfPodsToScaleDown; scaleDownIndex++ {
 		scaleDownPodName := podList.Items[wildflyServerNumberOfPods-scaleDownIndex].ObjectMeta.Name
 		wildflyServerSpecPodStatus := getWildflyServerPodStatusByName(w, scaleDownPodName)
@@ -845,12 +851,12 @@ func (r *ReconcileWildFlyServer) processTransactionRecoveryScaleDown(reqLogger l
 		if wildflyServerSpecPodStatus.State == wildflyv1alpha1.PodStateActive {
 			wildflyServerSpecPodStatus.State = wildflyv1alpha1.PodStateScalingDownRecoveryInvestigation
 			scaleDownPodsStates.Store(scaleDownPodName, wildflyv1alpha1.PodStateScalingDownRecoveryInvestigation)
-			updated = true
+			updated.Set()
 		} else {
 			scaleDownPodsStates.Store(scaleDownPodName, wildflyServerSpecPodStatus.State)
 		}
 	}
-	if updated { // updating status of pods as soon as possible
+	if updated.IsSet() { // updating status of pods as soon as possible
 		w.Status.ScalingdownPods = int32(numberOfPodsToScaleDown)
 		err := r.client.Status().Update(context.TODO(), w)
 		if err != nil {
@@ -859,6 +865,7 @@ func (r *ReconcileWildFlyServer) processTransactionRecoveryScaleDown(reqLogger l
 		return true, err
 	}
 
+	updated.UnSet()
 	var wg sync.WaitGroup
 	for scaleDownIndex := 1; scaleDownIndex <= numberOfPodsToScaleDown; scaleDownIndex++ {
 		scaleDownPod := podList.Items[wildflyServerNumberOfPods-scaleDownIndex]
@@ -874,14 +881,19 @@ func (r *ReconcileWildFlyServer) processTransactionRecoveryScaleDown(reqLogger l
 
 			podState, ok := scaleDownPodsStates.Load(scaleDownPodName)
 			if !ok {
-				scaleDownErrors.Store(scaleDownPodName,
+				scaleDownErrors.Store(scaleDownPodName+"_status-update",
 					fmt.Errorf("Cannot find pod name '%v' in the list of the active pods for the WildflyServer operator: %v",
 						scaleDownPodName, w.ObjectMeta.Name))
+				_, podsStatus := getPodStatus(podList.Items, w.Status.Pods)
+				reqLogger.Info("Updating pod status", "Pod Status", podsStatus)
+				w.Status.Pods = podsStatus
+				updated.Set()
 				return
 			}
 
 			if podState != wildflyv1alpha1.PodStateScalingDownClean {
-				reqLogger.Info("Transaction recovery scaledown processing", "Pod Name", scaleDownPodName, "IP Address", scaleDownPodIP)
+				reqLogger.Info("Transaction recovery scaledown processing", "Pod Name", scaleDownPodName,
+					"IP Address", scaleDownPodIP, "Pod State", podState, "Pod Phase", scaleDownPod.Status.Phase)
 
 				// TODO: check if it's possible to set graceful shutdown here
 				success, message, err := r.checkRecovery(reqLogger, &scaleDownPod, w)
@@ -906,11 +918,10 @@ func (r *ReconcileWildFlyServer) processTransactionRecoveryScaleDown(reqLogger l
 	wg.Wait()
 
 	// Updating the pod state based on the recovery processing when a scale down is in progress
-	updated = false
 	for wildflyServerPodStatusIndex, v := range w.Status.Pods {
 		if podStateValue, exist := scaleDownPodsStates.Load(v.Name); exist {
 			if w.Status.Pods[wildflyServerPodStatusIndex].State != podStateValue.(string) {
-				updated = true
+				updated.Set()
 			}
 			w.Status.Pods[wildflyServerPodStatusIndex].State = podStateValue.(string)
 		}
@@ -930,7 +941,7 @@ func (r *ReconcileWildFlyServer) processTransactionRecoveryScaleDown(reqLogger l
 			"Errors during transaction recovery scaledown processing. Consult operator log.")
 	}
 
-	if updated { // recovery changed the state of the pods
+	if updated.IsSet() { // recovery changed the state of the pods
 		w.Status.ScalingdownPods = int32(numberOfPodsToScaleDown)
 		err := r.client.Status().Update(context.TODO(), w)
 		if err != nil {
@@ -948,14 +959,18 @@ func (r *ReconcileWildFlyServer) checkRecovery(reqLogger logr.Logger, scaleDownP
 	scaleDownPodIP := scaleDownPod.Status.PodIP
 	scaleDownPodRecoveryPort := defaultRecoveryPort
 
+	// Pod is not started yet. For being safely marked as clean by recovery process it has to be fully operative
+	if scaleDownPod.Status.Phase == corev1.PodPending {
+		return false, "", fmt.Errorf("Pod '%s' / '%v' is in pending phase %v. It will be hopefully started in a while. "+
+			"Transaction recovery needs the pod being fully started to be capable to mark it as clean for the scale down.",
+			scaleDownPodName, w.ObjectMeta.Name, scaleDownPod.Status.Phase)
+	}
+
 	// Reading timestamp for the latest log record
 	scaleDownPodLogTimestampAtStart, err := wildflyutil.ObtainLogLatestTimestamp(scaleDownPod)
 	if err != nil {
-		if scaleDownPod.Status.Phase != corev1.PodRunning {
-			return false, "", fmt.Errorf("Pod '%s' of WildflyServer '%v' could not be ready for reading log during scaling down, "+
-				"please verify its state. Cause error: %v", scaleDownPodName, w.ObjectMeta.Name, err)
-		}
-		return false, "", fmt.Errorf("Failed to read log from scaling down pod '%v', error: %v", scaleDownPodName, err)
+		return false, "", fmt.Errorf("Log of the pod '%s' of WildflyServer '%v' is not ready during scaling down, "+
+			"please verify its state. Error: %v", scaleDownPodName, w.ObjectMeta.Name, err)
 	}
 
 	// If we are in state of recovery is needed the setup of the server has to be already done
@@ -991,8 +1006,10 @@ func (r *ReconcileWildFlyServer) checkRecovery(reqLogger logr.Logger, scaleDownP
 			reqLogger.Error(err, "Error on reading transaction recovery port with management command", "Pod name", scaleDownPodName)
 		}
 
-		// Marking the pod as being searched for the recovery port already
-		scaleDownPod.Annotations[markerRecoveryPort] = strconv.FormatInt(int64(scaleDownPodRecoveryPort), 10)
+		// The pod was already searched for the recovery port, marking that into the annotations
+		annotations := wildflyutil.MapMerge(
+			scaleDownPod.GetAnnotations(), map[string]string{markerRecoveryPort: strconv.FormatInt(int64(scaleDownPodRecoveryPort), 10)})
+		scaleDownPod.SetAnnotations(annotations)
 		if err := r.client.Update(context.TODO(), scaleDownPod); err != nil {
 			return false, "", fmt.Errorf("Failed to update pod annotations, pod name %v, annotations to be set %v, error: %v",
 				scaleDownPodName, scaleDownPod.Annotations, err)

--- a/pkg/controller/wildflyserver/wildflyserver_controller.go
+++ b/pkg/controller/wildflyserver/wildflyserver_controller.go
@@ -249,9 +249,9 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 			"Number of pods to be removed", numberOfPodsToScaleDown)
 	}
 
-	mustReconcile, requeue, err := r.checkStatefulSet(wildflyServer, foundStatefulSet, podList)
+	mustReconcile, err = r.checkStatefulSet(wildflyServer, foundStatefulSet, podList)
 	if mustReconcile {
-		return reconcile.Result{Requeue: requeue}, err
+		return reconcile.Result{Requeue: true}, err
 	} else if err != nil {
 		return reconcile.Result{}, err
 	}
@@ -365,10 +365,9 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 
 // checkStatefulSet checks if the statefulset is up to date with the current WildFlyServerSpec.
 // it returns true if a reconcile result must be returned.
-// the 2nd boolean specifies whether the result must be required.
 // A non-nil error if an error happend while updating/deleting the statefulset.
 func (r *ReconcileWildFlyServer) checkStatefulSet(wildflyServer *wildflyv1alpha1.WildFlyServer, foundStatefulSet *appsv1.StatefulSet,
-	podList *corev1.PodList) (mustReconcile bool, mustRequeue bool, err error) {
+	podList *corev1.PodList) (mustReconcile bool, err error) {
 	var update, requeue bool
 	// Ensure the statefulset replicas is up to date (driven by scaledown processing)
 	wildflyServerSpecSize := wildflyServer.Spec.Size
@@ -459,10 +458,10 @@ func (r *ReconcileWildFlyServer) checkStatefulSet(wildflyServer *wildflyv1alpha1
 					// VolumeClaimTemplates has changed, the statefulset can not be updated and must be deleted
 					if err = r.client.Delete(context.TODO(), foundStatefulSet); err != nil {
 						log.Error(err, "Failed to Delete StatefulSet.", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
-						return true, false, err
+						return true, err
 					}
 					log.Info("Deleting StatefulSet that is not up to date with the WildFlyServer StorageSpec", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
-					return true, true, nil
+					return true, nil
 				}
 
 				// all other changes are in the spec Template or Replicas and the statefulset can be updated
@@ -471,10 +470,10 @@ func (r *ReconcileWildFlyServer) checkStatefulSet(wildflyServer *wildflyv1alpha1
 				foundStatefulSet.Annotations["wildfly.org/wildfly-server-generation"] = strconv.FormatInt(wildflyServer.Generation, 10)
 				if err = r.client.Update(context.TODO(), foundStatefulSet); err != nil {
 					log.Error(err, "Failed to Update StatefulSet.", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
-					return true, false, err
+					return true, err
 				}
 				log.Info("Updating StatefulSet to be up to date with the WildFlyServer Spec", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
-				return true, true, nil
+				return true, nil
 			}
 		}
 	}
@@ -484,12 +483,12 @@ func (r *ReconcileWildFlyServer) checkStatefulSet(wildflyServer *wildflyv1alpha1
 		err := r.client.Update(context.TODO(), foundStatefulSet)
 		if err != nil {
 			log.Error(err, "Failed to update StatefulSet.", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
-			return false, true, err
+			return true, err
 		}
-		return false, true, nil
+		return true, nil
 	}
 
-	return false, requeue, nil
+	return requeue, nil
 }
 
 // matches checks if the envVar from the WildFlyServerSpec matches the same env var from the container.
@@ -1090,7 +1089,7 @@ func (r *ReconcileWildFlyServer) setupRecoveryPropertiesAndRestart(reqLogger log
 	if scaleDownPod.Annotations[markerRecoveryPropertiesSetup] == "" {
 		setBackoffPeriodOp := fmt.Sprintf(wildflyutil.MgmtOpSystemPropertyRecoveryBackoffPeriod, "1")
 		setOrphanIntervalOp := fmt.Sprintf(wildflyutil.MgmtOpSystemPropertyOrphanSafetyInterval, "1")
-		setRecoveryPeriodOp := fmt.Sprintf(wildflyutil.MgmtOpSystemPropertyPeriodicRecoveryPeriod, "1")
+		setRecoveryPeriodOp := fmt.Sprintf(wildflyutil.MgmtOpSystemPropertyPeriodicRecoveryPeriod, "30") // speed-up but still having time proceed the SCAN
 		wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, setBackoffPeriodOp)
 		wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, setOrphanIntervalOp)
 		jsonResult, err := wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, setRecoveryPeriodOp)

--- a/pkg/controller/wildflyserver/wildflyserver_controller.go
+++ b/pkg/controller/wildflyserver/wildflyserver_controller.go
@@ -1032,6 +1032,10 @@ func (r *ReconcileWildFlyServer) checkRecovery(reqLogger logr.Logger, scaleDownP
 	reqLogger.Info("Executing recovery scan at "+scaleDownPodName, "Pod IP", scaleDownPodIP, "Recovery port", scaleDownPodRecoveryPort)
 	_, err = wildflyutil.SocketConnect(scaleDownPodIP, scaleDownPodRecoveryPort, txnRecoveryScanCommand)
 	if err != nil {
+		delete(scaleDownPod.Annotations, markerRecoveryPort)
+		if err := r.client.Update(context.TODO(), scaleDownPod); err != nil {
+			reqLogger.Info("Cannot update scaledown pod %v while resetting the annotation map to %v", scaleDownPodName, scaleDownPod.Annotations)
+		}
 		return false, "", fmt.Errorf("Failed to run transaction recovery scan for scaling down pod %v. "+
 			"Please, verify the pod log file. Error: %v", scaleDownPodName, err)
 	}

--- a/pkg/controller/wildflyserver/wildflyserver_controller.go
+++ b/pkg/controller/wildflyserver/wildflyserver_controller.go
@@ -974,25 +974,27 @@ func (r *ReconcileWildFlyServer) checkRecovery(reqLogger logr.Logger, scaleDownP
 
 	// If we are in state of recovery is needed the setup of the server has to be already done
 	if scaleDownPod.Annotations[markerRecoveryPort] == "" {
-		reqLogger.Info("Enabling recovery listener for processing scaledown at " + scaleDownPodName)
-		// Enabling recovery listener to be able to call the server for the recovery
-		jsonResult, err := wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, wildflyutil.MgmtOpTxnEnableRecoveryListener)
+		reqLogger.Info("Verification the recovery listener is setup to run transaction recovery at " + scaleDownPodName)
+		// Verify the recovery listener is setup
+		jsonResult, err := wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, wildflyutil.MgmtOpTxnCheckRecoveryListener)
 		if err != nil {
 			if strings.Contains(strings.ToLower(err.Error()), "cannot execute") {
 				reqLogger.Error(err, "Verify if operator JBOSS_HOME variable determines the place where the application server is installed",
 					w.Name+".JBOSS_HOME", jbossHome)
 			}
-			return false, "", fmt.Errorf("Cannot enable transaction recovery listener for scaling down pod %v, error: %v", scaleDownPodName, err)
+			return false, "", fmt.Errorf("Cannot check if the transaction recovery listener is enabled for recovery at pod %v, error: %v", scaleDownPodName, err)
 		}
 		if !wildflyutil.IsMgmtOutcomeSuccesful(jsonResult) {
-			return false, "", fmt.Errorf("Failed to enable transaction recovery listener for scaling down pod %v. Scaledown processing cannot trigger recovery. "+
-				"Management command: %v, JSON response: %v", scaleDownPodName, wildflyutil.MgmtOpTxnEnableRecoveryListener, jsonResult)
+			return false, "", fmt.Errorf("Failed to verify if transaction recovery listener is enabled at pod %v. Scaledown processing cannot trigger recovery. "+
+				"Management command: %v, JSON response: %v", scaleDownPodName, wildflyutil.MgmtOpTxnCheckRecoveryListener, jsonResult)
 		}
-		// Enabling the recovery listner may require the server being reloaded
-		isReloadRequired := wildflyutil.ReadJSONDataByIndex(jsonResult["response-headers"], "operation-requires-reload")
-		if isReloadRequiredBool, _ := strconv.ParseBool(isReloadRequired); isReloadRequiredBool {
-			reqLogger.Info("Reloading WildFly for recovery listener being activated at pod " + scaleDownPodName)
-			wildflyutil.ExecuteOpAndWaitForServerBeingReady(wildflyutil.MgmtOpReload, scaleDownPod, jbossHome)
+		// When listener is not enabled then the pod will be terminated
+		if jsonResult["result"] != "true" {
+			reqLogger.Info("Transaction recovery listener is not enabled. Recovery pocess can't proceed at pod " + scaleDownPodName)
+			r.recorder.Event(w, corev1.EventTypeWarning, "WildFlyServerTransactionRecovery",
+				"Application server at pod "+scaleDownPodName+" does not define transaction recovery listener and recovery processing can't go forward."+
+					" Please consider fixing server configuration. The pod is now going to be terminated.")
+			return true, "", nil
 		}
 
 		// Reading recovery port from the app server with management port

--- a/pkg/controller/wildflyserver/wildflyserver_controller.go
+++ b/pkg/controller/wildflyserver/wildflyserver_controller.go
@@ -216,7 +216,7 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 	}
 
 	// List of pods which belongs under this WildflyServer instance
-	podList, err := r.getPodsForWildFly(wildflyServer)
+	podList, err := GetPodsForWildFly(r, wildflyServer)
 	if err != nil {
 		reqLogger.Error(err, "Failed to list pods.", "WildFlyServer.Namespace", wildflyServer.Namespace, "WildFlyServer.Name", wildflyServer.Name)
 		return reconcile.Result{}, err
@@ -509,11 +509,11 @@ func matches(container *v1.Container, envVar corev1.EnvVar) bool {
 	return false
 }
 
-// listing pods which belongs to the WildFly server
+// GetPodsForWildFly lists pods which belongs to the WildFly server
 //   the pods are differentiated based on the selectors
-func (r *ReconcileWildFlyServer) getPodsForWildFly(w *wildflyv1alpha1.WildFlyServer) (*corev1.PodList, error) {
+func GetPodsForWildFly(r *ReconcileWildFlyServer, w *wildflyv1alpha1.WildFlyServer) (*corev1.PodList, error) {
 	podList := &corev1.PodList{}
-	labelSelector := labels.SelectorFromSet(labelsForWildFly(w))
+	labelSelector := labels.SelectorFromSet(LabelsForWildFly(w))
 	listOps := &client.ListOptions{
 		Namespace:     w.Namespace,
 		LabelSelector: labelSelector,
@@ -529,7 +529,7 @@ func (r *ReconcileWildFlyServer) getPodsForWildFly(w *wildflyv1alpha1.WildFlySer
 
 // statefulSetForWildFly returns a wildfly StatefulSet object
 func (r *ReconcileWildFlyServer) statefulSetForWildFly(w *wildflyv1alpha1.WildFlyServer) *appsv1.StatefulSet {
-	ls := labelsForWildFly(w)
+	ls := LabelsForWildFly(w)
 	// track the generation number of the WildFlyServer that created the statefulset to ensure that the
 	// statefulset is always up to date with the WildFlyServerSpec
 	annotations := make(map[string]string)
@@ -538,7 +538,7 @@ func (r *ReconcileWildFlyServer) statefulSetForWildFly(w *wildflyv1alpha1.WildFl
 	replicas := w.Spec.Size
 	applicationImage := w.Spec.ApplicationImage
 	volumeName := w.Name + "-volume"
-	labesForActiveWildflyPod := labelsForWildFly(w)
+	labesForActiveWildflyPod := LabelsForWildFly(w)
 	labesForActiveWildflyPod[markerOperatedByHeadless] = markerServiceActive
 	labesForActiveWildflyPod[markerOperatedByLoadbalancer] = markerServiceActive
 
@@ -671,7 +671,7 @@ func (r *ReconcileWildFlyServer) statefulSetForWildFly(w *wildflyv1alpha1.WildFl
 
 // loadBalancerForWildFly returns a loadBalancer service
 func (r *ReconcileWildFlyServer) loadBalancerForWildFly(w *wildflyv1alpha1.WildFlyServer) *corev1.Service {
-	labels := labelsForWildFly(w)
+	labels := LabelsForWildFly(w)
 	labels[markerOperatedByLoadbalancer] = markerServiceActive // managing only active pods which are not in scaledown process
 	sessionAffinity := corev1.ServiceAffinityNone
 	if w.Spec.SessionAffinity {
@@ -702,7 +702,7 @@ func (r *ReconcileWildFlyServer) loadBalancerForWildFly(w *wildflyv1alpha1.WildF
 
 // headlessServiceForWildFly returns a headless service for ejb remoting server to server calls
 func (r *ReconcileWildFlyServer) headlessServiceForWildFly(w *wildflyv1alpha1.WildFlyServer) *corev1.Service {
-	labels := labelsForWildFly(w)
+	labels := LabelsForWildFly(w)
 	labels[markerOperatedByHeadless] = markerServiceActive // managing only active pods, ones are not scaled down
 	headlessService := &corev1.Service{
 		ObjectMeta: metav1.ObjectMeta{
@@ -738,7 +738,7 @@ func (r *ReconcileWildFlyServer) routeForWildFly(w *wildflyv1alpha1.WildFlyServe
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      w.Name,
 			Namespace: w.Namespace,
-			Labels:    labelsForWildFly(w),
+			Labels:    LabelsForWildFly(w),
 		},
 		Spec: routev1.RouteSpec{
 			To: routev1.RouteTargetReference{
@@ -758,7 +758,7 @@ func (r *ReconcileWildFlyServer) routeForWildFly(w *wildflyv1alpha1.WildFlyServe
 }
 
 func (r *ReconcileWildFlyServer) finalizeWildflyServer(reqLogger logr.Logger, w *wildflyv1alpha1.WildFlyServer) (bool, error) {
-	podList, err := r.getPodsForWildFly(w)
+	podList, err := GetPodsForWildFly(r, w)
 	if err != nil {
 		return false, fmt.Errorf("Finalizer processing: failed to list pods for WildflyServer %v:%v name Error: %v", w.Namespace, w.Name, err)
 	}
@@ -1231,7 +1231,9 @@ func envForClustering(labels string) []corev1.EnvVar {
 	}
 }
 
-func labelsForWildFly(w *wildflyv1alpha1.WildFlyServer) map[string]string {
+// LabelsForWildFly retirms lists of labels that are used for identification
+//  of objects belonging to the particular WildflyServer instance
+func LabelsForWildFly(w *wildflyv1alpha1.WildFlyServer) map[string]string {
 	labels := make(map[string]string)
 	labels["app.kubernetes.io/name"] = w.Name
 	labels["app.kubernetes.io/managed-by"] = os.Getenv("LABEL_APP_MANAGED_BY")

--- a/pkg/controller/wildflyserver/wildflyserver_controller_test.go
+++ b/pkg/controller/wildflyserver/wildflyserver_controller_test.go
@@ -3,11 +3,13 @@ package wildflyserver
 import (
 	"context"
 	"testing"
+	"time"
 
 	wildflyv1alpha1 "github.com/wildfly/wildfly-operator/pkg/apis/wildfly/v1alpha1"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -223,4 +225,116 @@ func TestEnvUpdate(t *testing.T) {
 			assert.Equal("ADD", env.Value)
 		}
 	}
+}
+
+func TestWildFlyServerControllerScaleDown(t *testing.T) {
+	// Set the logger to development mode for verbose logs.
+	logf.SetLogger(logf.ZapLogger(true))
+	assert := assert.New(t)
+	expectedReplicaSize := int32(1)
+
+	// A WildFlyServer resource with metadata and spec.
+	wildflyServer := &wildflyv1alpha1.WildFlyServer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: wildflyv1alpha1.WildFlyServerSpec{
+			ApplicationImage: applicationImage,
+			Size:             expectedReplicaSize,
+			SessionAffinity:  sessionAffinity,
+		},
+	}
+	// Objects to track in the fake client.
+	objs := []runtime.Object{
+		wildflyServer,
+	}
+
+	// Register operator types with the runtime scheme.
+	s := scheme.Scheme
+	s.AddKnownTypes(wildflyv1alpha1.SchemeGroupVersion, wildflyServer)
+	// Create a fake client to mock API calls.
+	cl := fake.NewFakeClient(objs...)
+	// Create a ReconcileWildFlyServer object with the scheme and fake client.
+	r := &ReconcileWildFlyServer{client: cl, scheme: s, recorder: eventRecorderMock{}, isOpenShift: false, isWildFlyFinalizer: true}
+	// Mock request to simulate Reconcile() being called on an event for a
+	// watched resource .
+	req := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	// Statefulset will be created
+	_, err := r.Reconcile(req)
+	require.NoError(t, err)
+
+	// Check if stateful set has been created and has the correct size.
+	statefulSet := &appsv1.StatefulSet{}
+	err = cl.Get(context.TODO(), req.NamespacedName, statefulSet)
+	require.NoError(t, err)
+	assert.Equal(expectedReplicaSize, *statefulSet.Spec.Replicas)
+	assert.Equal(applicationImage, statefulSet.Spec.Template.Spec.Containers[0].Image)
+
+	// Finalizer will be added
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+	err = cl.Get(context.TODO(), req.NamespacedName, wildflyServer)
+	require.NoError(t, err)
+	assert.Equal(1, len(wildflyServer.GetFinalizers()))
+	assert.Equal("finalizer.wildfly.org", wildflyServer.GetFinalizers()[0])
+	// Operator correctly setup the StatefulSet replica thus move on and create the Pod that the operator waits for
+	//   StatefulSet won't do this here for us thus manual creation is needed
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{Name: "ScaleDownTestPod", Namespace: wildflyServer.Namespace, Labels: LabelsForWildFly(wildflyServer)},
+		TypeMeta:   metav1.TypeMeta{Kind: "Pod", APIVersion: "v1"}}
+	err = cl.Create(context.TODO(), pod)
+	require.NoError(t, err)
+
+	log.Info("Waiting for WildflyServer is updated to the state where WildflyServer.Status.Pods refers the Pod created by the test")
+	err = wait.Poll(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+		_, err = r.Reconcile(req)
+		require.NoError(t, err)
+
+		podList, err := GetPodsForWildFly(r, wildflyServer)
+		err2 := cl.Get(context.TODO(), req.NamespacedName, wildflyServer)
+		if err == nil && len(podList.Items) == int(expectedReplicaSize) &&
+			err2 == nil && len(wildflyServer.Status.Pods) == int(expectedReplicaSize) {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	log.Info("WildFly server was reconciliated to the state the pod status corresponds with namespace. Let's scale it down.",
+		"WildflyServer", wildflyServer)
+	assert.Equal(int(expectedReplicaSize), len(wildflyServer.Status.Pods))
+	assert.Equal(wildflyv1alpha1.PodStateActive, wildflyServer.Status.Pods[0].State)
+	wildflyServer.Spec.Size = 0
+	err = cl.Update(context.TODO(), wildflyServer)
+
+	// Reconcile for the scale down - updating the pod labels
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+	// Pod label has to be changed to not being active for service
+	podList, err := GetPodsForWildFly(r, wildflyServer)
+	require.NoError(t, err)
+	assert.Equal(int(expectedReplicaSize), len(podList.Items))
+	assert.Equal("disabled", podList.Items[0].GetLabels()["wildfly.org/operated-by-loadbalancer"])
+
+	// Reconcile for the scale down - updating the pod state at the wildflyserver CR
+	_, err = r.Reconcile(req) // error could be returned here as the scaledown was not sucessful here
+	err = cl.Get(context.TODO(), req.NamespacedName, wildflyServer)
+	require.NoError(t, err)
+	assert.Equal(wildflyv1alpha1.PodStateScalingDownRecoveryInvestigation, wildflyServer.Status.Pods[0].State)
+}
+
+type eventRecorderMock struct {
+}
+
+func (rm eventRecorderMock) Event(object runtime.Object, eventtype, reason, message string) {}
+func (rm eventRecorderMock) Eventf(object runtime.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+func (rm eventRecorderMock) PastEventf(object runtime.Object, timestamp metav1.Time, eventtype, reason, messageFmt string, args ...interface{}) {
+}
+func (rm eventRecorderMock) AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{}) {
 }

